scipy/linalg/_solvers.py

```diff
@@ -329,44 +343,81 @@ def solve_discrete_are(a, b, q, r):
     Notes
     -----
     Method taken from:
-    Laub, "A Schur Method for Solving Algebraic Riccati Equations."
-    U.S. Energy Research and Development Agency under contract
-    ERDA-E(49-18)-2087.
-    http://dspace.mit.edu/bitstream/handle/1721.1/1301/R-0859-05666488.pdf
+    P. van Dooren , "A Generalized Eigenvalue Approach For Solving Riccati 
+    Equations.", SIAM Journal on Scientific and Statistical Computing,
+    Vol.2(2), DOI: 10.1137/0902010
+
 
     .. versionadded:: 0.11.0
 
     """
-
-    try:
-        g = inv(r)
-    except LinAlgError:
-        raise ValueError('Matrix R in the algebraic Riccati equation solver '
-                         'is ill-conditioned')
-
-    g = np.dot(np.dot(b, g), b.conj().transpose())
-
-    try:
-        ait = inv(a).conj().transpose()  # ait is "A inverse transpose"
-    except LinAlgError:
-        raise ValueError('Matrix A in the algebraic Riccati equation solver '
-                         'is ill-conditioned')
-
-    z11 = a+np.dot(np.dot(g, ait), q)
-    z12 = -1.0*np.dot(g, ait)
-    z21 = -1.0*np.dot(ait, q)
-    z22 = ait
-
-    z = np.vstack((np.hstack((z11, z12)), np.hstack((z21, z22))))
-
-    # Note: we need to sort the upper left of s to lie within the unit circle,
-    #       while the lower right is outside (Laub, p. 7)
-    s, u, _ = schur(z, sort='iuc')
-
-    (m, n) = u.shape
-
-    u11 = u[0:m//2, 0:n//2]
-    u21 = u[m//2:m, 0:n//2]
-    u11i = inv(u11)
-
-    return np.dot(u21, u11i)
+    
+    a = _asarray_validated(a)
+    b = _asarray_validated(b)
+    q = _asarray_validated(q)
+    r = _asarray_validated(r)
+    
+    # Get the correct data types otherwise Numpy complains about pushing 
+    # complex numbers into real arrays.
+    r_or_c = complex if np.iscomplexobj(b) else float
+
+    for ind, x in enumerate((a,q,r)):
+        if np.iscomplexobj(x):
+            r_or_c = complex
+        if not np.equal(*x.shape):
+            raise ValueError("Matrix {} should be square.".format("aqr"[ind]))
+    
+    # Shape consistency checks
+    m, n = b.shape
+
+    if m != a.shape[0]:
+        raise ValueError("Matrix a and b should have the same number of rows.")
+        
+    if m != q.shape[0]:
+        raise ValueError("Matrix a and q should have the same shape.")
+            
+    if n != r.shape[0]:
+        raise ValueError("Matrix b and r should have the same number of cols.")
+    
+    # Check if the data is (sufficiently) hermitian
```

Comment:
If we don't do it, it will trip somewhere in the code with the cryptic broadcast exception. I think it is better to show why it fails such that the user immediately spots the mistake.